pkg/zfs/plugin/backup.go

```diff
@@ -144,34 +145,44 @@ func (p *Plugin) createBackup(vol *apis.ZFSVolume, schdname, snapname string, po
 		return "", err
 	}
 
-	return bkpname, err
+	return bkpname, nil
 }
 
-func (p *Plugin) checkBackupStatus(bkpname string) {
-	bkpDone := false
-
-	for !bkpDone {
+func (p *Plugin) checkBackupStatus(bkpname string) error {
+	for {
 		getOptions := metav1.GetOptions{}
 		bkp, err := bkpbuilder.NewKubeclient().
 			WithNamespace(p.namespace).Get(bkpname, getOptions)
 
 		if err != nil {
 			p.Log.Errorf("zfs: Failed to fetch backup info {%s}", bkpname)
-			p.cl.ExitServer = true
-			return
+			return errors.Errorf("zfs: error in getting bkp status err %v", err)
 		}
 
-		time.Sleep(backupStatusInterval * time.Second)
-
 		switch bkp.Status {
-		case apis.BKPZFSStatusDone, apis.BKPZFSStatusFailed, apis.BKPZFSStatusInvalid:
-			bkpDone = true
-			p.cl.ExitServer = true
+		case apis.BKPZFSStatusDone:
+			return nil
+		case apis.BKPZFSStatusFailed, apis.BKPZFSStatusInvalid:
+			return errors.Errorf("zfs: error in uploading snapshot, status:{%v}", bkp.Status)
 		}
+
+		time.Sleep(backupStatusInterval * time.Second)
+	}
+}
+
+func (p *Plugin) doUpload(wg *sync.WaitGroup, filename string, size int64, port int) {
+	defer wg.Done()
+
+	ok := p.cl.Upload(filename, size, port)
+	if !ok {
+		p.Log.Errorf("zfs: Failed to upload file %s", filename)
+		*p.cl.ConnReady <- false
 	}
+	// done with the channel, close it
+	close(*p.cl.ConnReady)
```

Comment:
Should we move this to `clouduploader.Run` and, initialization in `clouduploader.Init`? So if caller wants to wait for the server status then caller can use `clouduploader.ConnReadyWait` else can skip it..  This way, caller don't need to initialize or handle this channel in upload/download go-routine.

Reply:
hmmm, good suggestion.

Reply:
ok, I remember why I did this way, this waitgroup has to be passed when we want to run it in a go routine. Because of that I did it this way.

Reply:
>This way, caller don't need to initialize or handle this channel in upload/download go-routine.
For backing multiple volumes, Init gets called only once. So that might not be the right place to intialize it. Here we are initializing it when ever we are starting the upload/download go routine to make sure a new dedicated channel gets used for each upload/download and we are making sure that the channel get closed also.